### ISSUE TITLE
Generic class parameter not being bound to query [DATACMNS-1002]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-commons#1452]
**[Michael Bragg](https://jira.spring.io/secure/ViewProfile.jspa?name=mbragg)** opened **[DATACMNS-1002](https://jira.spring.io/browse/DATACMNS-1002?redirect=false)** and commented

Im in the process of upgrading our app and I've run into an issue related to changes made [here](https://github.com/spring-projects/spring-data-commons/pull/150) to `Parameter`. We are upgrading from Spring Data Commons from 1.11.4.RELEASE to 1.12.x / 1.13.x. We have a hierarchical entity structure. e.g. `ReasonLabel extends AbstractResolutionLabel`. We have a table with a Discriminator Column for the name of the concrete type. We have the following query which previously worked:

```java
@Query("SELECT label FROM AbstractResolutionLabel label WHERE label.group.id = :groupId AND label.providerType = :providerType AND TYPE(label) = :clazz")
<T extends AbstractResolutionLabel> List<T> findByGroupIdAndProviderTypeAndType(@Param("groupId") String groupId,
  @Param("providerType") ProviderType providerType, @Param("clazz") Class<T> clazz);
```

We are now getting the following error:

```
Caused by: org.hibernate.QueryException: Named parameter [clazz] not set
	at org.hibernate.query.internal.QueryParameterBindingsImpl.verifyParametersBound(QueryParameterBindingsImpl.java:234)
	at org.hibernate.query.internal.AbstractProducedQuery.beforeQuery(AbstractProducedQuery.java:1309)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1408)
	at org.hibernate.query.Query.getResultList(Query.java:146)
	at org.springframework.data.jpa.repository.query.JpaQueryExecution$CollectionExecution.doExecute(JpaQueryExecution.java:121)
	at org.springframework.data.jpa.repository.query.JpaQueryExecution.execute(JpaQueryExecution.java:85)
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.doExecute(AbstractJpaQuery.java:116)
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.execute(AbstractJpaQuery.java:106)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:483)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:461)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
	... 57 more
```

It appears that from [this line](https://github.com/spring-projects/spring-data-commons/pull/150/files#diff-325a1f477d275feb68e9ffec76dc064eR200) the new `isDynamicProjectionParameter(…)` is returning `true`, so the class parameter is not being bound, because both the parameter type and return type (T) are both equal.

I appreciate this change has been here for some time now. This seems like a valid query that should be work. Many thanks

---

**Affects:** 1.13.1 (Ingalls SR1), 1.12.8 (Hopper SR8)

**Reference URL:** https://github.com/spring-projects/spring-data-commons/pull/150

**Referenced from:** pull request https://github.com/spring-projects/spring-data-commons/pull/199

**Comment by rafg92** [comment]
Hi, are there any updates about this issue?

**Comment by mp911de** [comment]
Fixed via #2770
